### PR TITLE
Dispatch events on Two Factor Authentication Success or Failure action

### PR DESCRIPTION
**Pull request by pmontoya**
Due to some git flow mistakes and some developpement loked in my company, I create a new pull request from my repository (and not BLEUROY-HIGHCO)

I create TwoFactorAuthenticationEvents and TwoFactorAuthenticationEvent as asked in previous PR but I think that my implementation was more faithful with Symfony doc

### CODE REVIEW

File: Resources/config/security.xml [hunk 1]
```diff
@@ -41,7 +41,9 @@
 		</service>
 		<service id="scheb_two_factor.provider_registry" class="%scheb_two_factor.provider_registry.class%" lazy="true">
 			<argument type="service" id="scheb_two_factor.session_flag_manager" />
-			<argument /> <!-- Two-factor providers -->
+            <argument type="service" id="event_dispatcher" />
+            <argument type="string" id="%auth_code%" />
+            <argument /> <!-- Two-factor providers -->
```
Comment (zerkms) [inline Resources/config/security.xml]: If you shift the providers argument to another position - you need to change the compiler pass that relies on it https://github.com/scheb/two-factor-bundle/blob/master/DependencyInjection/Compiler/ProviderCompilerPass.php#L36

Comment (pmontoya) [inline Resources/config/security.xml]: Sorry. I made this change in local but it wasn't pushed. Done.

File: Resources/config/security.xml [hunk 2]
```diff
@@ -41,7 +41,9 @@
 		</service>
 		<service id="scheb_two_factor.provider_registry" class="%scheb_two_factor.provider_registry.class%" lazy="true">
 			<argument type="service" id="scheb_two_factor.session_flag_manager" />
-			<argument /> <!-- Two-factor providers -->
+            <argument type="service" id="event_dispatcher" />
+            <argument type="string" id="%auth_code%" />
```
Comment (scheb) [inline Resources/config/security.xml]: @pmontoya That line must be, can you please change?
`<argument>%scheb_two_factor.parameter_names.auth_code%</argument>`

File: Security/TwoFactor/Provider/TwoFactorProviderRegistry.php [hunk 3]
```diff
@@ -70,7 +91,12 @@ public function requestAuthenticationCode(AuthenticationContextInterface $contex
 
                 // Set authentication completed
                 if ($context->isAuthenticated()) {
+                    $this->eventDispatcher->dispatch(TwoFactorAuthenticationEvents::SUCCESS, new TwoFactorAuthenticationEvent());
                     $this->flagManager->setComplete($providerName, $token);
+                } else {
+                    if ($context->getRequest()->request->has($this->authRequestParameter)) {
```
Comment (scheb) [inline Security/TwoFactor/Provider/TwoFactorProviderRegistry.php]: Please merge that "else" and "if" to an "else if" statement.

**Comment by scheb** [comment]
Two more changes (see above), then we can merge.

**Comment by pmontoya** [comment]
@scheb : done

**Pull request merged**